File: app/src/main/java/org/mifos/selfserviceapp/utils/QrCodeGenerator.java
package org.mifos.selfserviceapp.utils;

import android.graphics.Bitmap;

import com.google.gson.Gson;
import com.google.zxing.BarcodeFormat;
import com.google.zxing.MultiFormatWriter;
import com.google.zxing.WriterException;
import com.google.zxing.common.BitMatrix;

import org.mifos.selfserviceapp.models.beneficary.Beneficiary;
import org.mifos.selfserviceapp.models.templates.account.AccountType;

import static android.graphics.Color.BLACK;
import static android.graphics.Color.WHITE;

/**
 * Created by dilpreet on 6/7/17.
 */

public class QrCodeGenerator {

    public static Bitmap encodeAsBitmap(String str) {
        BitMatrix result;
        try {
            result = new MultiFormatWriter().encode(str, BarcodeFormat.QR_CODE, 200, 200, null);
        } catch (IllegalArgumentException iae) {
            return null;
        } catch (WriterException e) {
            return null;
        }

        int width = result.getWidth();
        int height = result.getHeight();
        int[] pixels = new int[width * height];
        for (int y = 0; y < height; y++) {
            int offset = y * width;
            for (int x = 0; x < width; x++) {
                pixels[offset + x] = result.get(x, y) ? BLACK : WHITE;
            }
        }

        Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        bitmap.setPixels(pixels, 0, width, 0, 0, width, height);
        return bitmap;
    }

    public static String getAccountDetailsInString(String accountNumber, String officeName,
                                                   org.mifos.selfserviceapp.ui.enums.AccountType
                                                           accountType) {
        int accountId = -1;
        if (accountType == org.mifos.selfserviceapp.ui.enums.AccountType.SAVINGS) {
            accountId = 0;
        } else if (accountType == org.mifos.selfserviceapp.ui.enums.AccountType.LOAN) {
            accountId = 1;
        }

        AccountType type = new AccountType();
        type.setId(accountId);
        Beneficiary payload = new Beneficiary();
        payload.setAccountNumber(accountNumber);
        payload.setAccountType(type);
        payload.setOfficeName(officeName);

        Gson gson = new Gson();
        return gson.toJson(payload);
    }
}

File: core/qrcode/src/main/kotlin/org/mifos/mobile/core/qr/QrCodeGenerator.kt
/*
 * Copyright 2024 Mifos Initiative
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See https://github.com/openMF/mobile-mobile/blob/master/LICENSE.md
 */
package org.mifos.mobile.core.qr

import android.graphics.Bitmap
import android.graphics.Color
import com.google.gson.Gson
import com.google.zxing.BarcodeFormat
import com.google.zxing.MultiFormatWriter
import com.google.zxing.common.BitMatrix
import org.mifos.mobile.core.model.entity.beneficiary.Beneficiary
import org.mifos.mobile.core.model.entity.templates.account.AccountType
import org.mifos.mobile.core.model.enums.AccountType as EnumsAccountType

object QrCodeGenerator {
    private const val QR_CODE_SIZE = 200

    /**
     * Generate a QRCode which stores `str` in the form of [Bitmap]
     * @param str Data which need to stored in QRCode
     * @return Returns a [Bitmap] of QRCode or null if generation fails
     */
    fun encodeAsBitmap(str: String?): Bitmap? {
        if (str.isNullOrEmpty()) return null

        return try {
            val result = MultiFormatWriter().encode(
                str,
                BarcodeFormat.QR_CODE,
                QR_CODE_SIZE,
                QR_CODE_SIZE,
                null,
            )
            createBitmapFromBitMatrix(result)
        } catch (e: Exception) {
            null
        }
    }

    private fun createBitmapFromBitMatrix(matrix: BitMatrix): Bitmap {
        val width = matrix.width
        val height = matrix.height
        val pixels = IntArray(width * height) { index ->
            if (matrix[index % width, index / width]) Color.BLACK else Color.WHITE
        }

        return Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888).apply {
            setPixels(pixels, 0, width, 0, 0, width, height)
        }
    }

    /**
     * Provides a string which contains json data for creating a [Beneficiary]
     * @param accountNumber Account Number of client
     * @param officeName Office Name of Client
     * @param accountType [EnumsAccountType] i.e. SAVINGS or LOAN
     * @return Returns a string with account details
     */
    fun getAccountDetailsInString(
        accountNumber: String?,
        officeName: String?,
        accountType: EnumsAccountType,
    ): String {
        val payload = Beneficiary().apply {
            this.accountNumber = accountNumber
            this.accountType = AccountType().apply {
                id = when (accountType) {
                    EnumsAccountType.SAVINGS -> 0
                    EnumsAccountType.LOAN -> 1
                    EnumsAccountType.SHARE -> -1
                }
            }
            this.officeName = officeName
        }

        return Gson().toJson(payload)
    }
}
